CONTENT: two new blogs on package health (#116)
pyOpenSci/pyopensci.github.io: _posts/2022-11-17-what-makes-a-python-package-healthy.md

---
layout: single
title: "What Makes a Python Open Source Package Healthy? A Conversation on Twitter."
excerpt: "How should pyOpenSci measure Python open source package health and level of maintenance and usability? Here I summarize a conversation held on twitter around this very topic. Feedback is welcome!"
author: "Leah Wasser"
permalink: /blog/what-makes-open-source-python-package-healthy
header:
    overlay_color: "#666"
    overlay_filter: 0.6
categories:
  - blog-post
  - highlight
  - python-packaging
  - peer-review
toc: true
comments: true
---

In previous posts I discussed:

* [Why Python open source software matters to scientists (and package maintainers too)]()
* [Why collecting metrics surrounding package health matters]()

In this post, I'll summarize a conversation that was held on twitter that gaged 
what the community thought about metrics to track the health of scientific 
Python open source packages. 

## Packages and open source software, a few terms to clarify

> * When I say Python package or Python open source software, i'm referring to a 

Reviewer: ```suggestion
* When I say Python package or Python open source software, I'm referring to a 
```

capitalized "I'm"

Reviewer: :thinking: this sounds a bit repetitive, specifically the use of "open source" in both sentences. Maybe:

> * When I say Python package, I'm referring to software anyone can install, for example from PyPI or off of GitHub. 
> * When I say "open source" or "free and open source software" I'm referring to code that is free to download where the source code itself is openly available for anyone to see.

tool that anyone can install to use in their Python environment. 

> * When i saw open source or free and open source software i'm referring to Python tools that are free to download and and their code is openly available for anyone to see (open source).

Reviewer: ```suggestion
* When I say open source or free and open source software I'm referring to Python tools that are free to download and and their code is openly available for anyone to see (open source).
```

## What is package health anyway? 

There are many different ways to think about and evaluate open source 
Python package health. 

Below is what I posted on Twitter to spur a conversation about what makes a 

> package health. And more specifically what metrics should we (pyOpenSci) collect 

Reviewer: ```suggestion
package healthy. And more specifically what metrics should we (pyOpenSci) collect 
```
"package health" -> "package healthy"

<script async src="https://platform.twitter.com/widgets.js" charset="utf-8"></script>

The twitter convo made me realize that there are
many different perspectives that we can consider when addressing this question. 

More specifically, pyOpenSci is interested in the health of packages that
support science. So we may need to build upon already existing 

> efforts that determine metrics and customize them to our needs.

Reviewer: Should "determine metrics" be something like perfect past tense here?  
> Existing efforts that have determined what metrics to use for package health

Maybe this is a way people that discuss metrics talk and it's just my ear that's not familiar with it

<div class="notice" markdown="1">
### A note about our pyOpenSci packages 

> Packages in the pyOpenSci ecosystem are not the large packages like Xarray, 

Reviewer: Would it be helpful to show the figure from Jake Vanderplas' 2017 pycon talk here
https://www.youtube.com/watch?v=ZyjCqQEUa8o

As referenced here: https://www.coiled.io/blog/pydata-dask

And maybe talk about packages that are "higher up in the stack" or "foundational"?

Author: that's a great idea!! perfect image for this. we'd kind of start where astropy is altho astropy is an ecosystem so we are kind of like an intermediate between astropy and scikit  :) i almost need to make a graphic for this i'd use it in presentations too.

Numpy of the SciPy ecosystem. Rather, they are often smaller and often developed 
by scientists for domain specific applications that others in their
community require. Our package maintainers:

* Often don't have the resources to build community

> * Often are keen to build users of their tools and to contribute

Reviewer: ```suggestion
* Often are keen to build their user base and to contribute
```
"build users" doesn't seem quite right (although I guess I would just manufacture users, maybe they'd star my repos more than my flesh-and-blood users :sob: )

The Software Sustainability Institute lead by Neil P Chue Hong has also 
thought about package health extensively and pulled together some data accordingly. Neil was also
a critical guiding member of the earlier pyOpenSci community meetings that were 
held in 2018.

<blockquote class="twitter-tweet"><p lang="en" dir="ltr">We also did some initial work on this in 2017 (see slide 12 of this presentation): <a href="https://t.co/1F0iMwfT5g">https://t.co/1F0iMwfT5g</a></p>&mdash; Neil P Chue Hong (he/him｜they/them) (@npch) <a href="https://twitter.com/npch/status/1577959663408218114?ref_src=twsrc%5Etfw">October 6, 2022</a></blockquote> <script async src="https://platform.twitter.com/widgets.js" charset="utf-8"></script> 

### Snyk and security (which aren't discussed in this post)

> One topic that I am not delving into in this post is security issues. SNYK 

Reviewer: ```suggestion
One topic that I am not delving into in this post is security issues. Snyk 
```

> is definitely a leader in this space and was mentioned atleast once in 

Reviewer: ```suggestion
is definitely a leader in this space and was mentioned at least once in 
```

the conversation. 

### Some existing metric examples 

> Below are some of the metrics that you can easily access via SNYK's website. 

Reviewer: ```suggestion
Below are some of the metrics that you can easily access via Snyk's website. 
```

Pretty sure Snyk writes their name as a proper noun, capitalized. That's how it shows up in the menu bar when I go to snyk.io and that's how it's written in text on the landing page

<blockquote class="twitter-tweet"><p lang="en" dir="ltr">This might be helpful. This website collects various metrics. And here is the example for numpy. <a href="https://t.co/YNsRoMgks4">https://t.co/YNsRoMgks4</a></p>&mdash; Kevin Wang (@KevinWangStats) <a href="https://twitter.com/KevinWangStats/status/1577999823911161856?ref_src=twsrc%5Etfw">October 6, 2022</a></blockquote> <script async src="https://platform.twitter.com/widgets.js" charset="utf-8"></script> 

#### [Pandera python package metrics on Snyk](https://snyk.io/advisor/python/pandera)

<figure>
    <a href="/images/pyopensci-packages/snyk-pandera-python-package-metrics.png"><img src="/images/pyopensci-packages/snyk-pandera-python-package-metrics.png" alt="Image showing the metrics from the Snyk website for the `pandera` package in our ecosystem."></a>

>     <figcaption>Here you can see what a snyk report looks like for pandera, a package accepted into our ecosystem a few years ago. Pandera gets a very healthy report because it's heavily used among other metrics.

Reviewer: ```suggestion
    <figcaption>Here you can see what a Snyk report looks like for pandera, a package accepted into our ecosystem a few years ago. Pandera gets a very healthy report because it's heavily used among other metrics.
```

## What metrics should pyOpenSci track for their Python scientific open source packages?

So back to the question at hand, what should pyOpenSci be tracking for [packages 
in our ecosystem](/python-packages/#our-accepted-python-open-source-packages)?
Hao Ye (and a few others) nailed it - health metrics are multi-dimensional. 

<blockquote class="twitter-tweet"><p lang="en" dir="ltr">I think, much like &quot;ecological stability&quot; - <a href="https://t.co/lJe2Fa0ycR">https://t.co/lJe2Fa0ycR</a> - , &quot;health&quot; here is multi-dimensional and different metrics will capture different facets, such as growth, transparency in governance, stability / backwards compatibility, etc.</p>&mdash; Hao Ye will haunt you for bad keming (@Hao_and_Y) <a href="https://twitter.com/Hao_and_Y/status/1577739685841059842?ref_src=twsrc%5Etfw">October 5, 2022</a></blockquote> <script async src="https://platform.twitter.com/widgets.js" charset="utf-8"></script> 

> I may be a bit biased here considering I have a degree in ecology BUT... I definitely support the ecological perspective always and forever :)

> As Justin from Napari and CZI points out, metrics are also perspective based. 

Reviewer: ```suggestion
As Justin Kiggins from Napari and CZI points out, metrics are also perspective based. 
```

Alas it is true that metrics designed for reporting that a funder requires
for a grant may differ metrics designed fom internal evaluation that 
informs program development. pyOpenSci has a lot to unpack there over the 
upcoming months!

### Three open source software healthy metric "buckets" 

Based on all of the Twitter feedback (below), and what I *think* might be a 
start at what pyOpenSci needs, I organized the Twitter conversation into three 
buckets:

1. Infrastructure
2. Maintenance 
3. Community adoption (and usability??)

These three buckets are all priorities of pyOpenSci.

<i class="far fa-sticky-note"></i> DEIA is another critical concern for pyOpenSci but I won't discuss that in this blog post. 
{: .notice }

## Infrastructure in a Python open source GitHub repository as a measure of package health

So here I start with Python package infrastructure found in a GitHub repository 
as a preliminary measure of package health. 
When think of infrastructure I think about the files and "things" available in a 

> repository that support it's use. I know that no bucket is perfectly isolated

Reviewer: ```suggestion
repository that support its use. I know that no bucket is perfectly isolated
```
no apostrophe in possessive its

from the others but i'm taking a stab at this here. 

> <i class="fas fa-info-circle"></i> The code for many open source software packages can be found on [GitHub](https://www.github.com). GitHub is a free-to-use website that runs `git` which is a version control system. Version control allows developers to track historical changes to code and files. GitHub which is a website allows developers to communicate openly, review new code changes and update content in a structured way. 

Reviewer: ```suggestion
<i class="fas fa-info-circle"></i> The code for many open source software packages can be found on [GitHub](https://www.github.com). GitHub is a free-to-use website that runs `git` which is a version control system. Version control allows developers to track historical changes to code and files. As a platform built on top of git, GitHub allows developers to communicate openly, review new code changes and update content in a structured way. 
```

{: .notice }

### What does GitHub (and Ivan) think about health checks for Python open source software?

> Ivan, is a long time advisor, editor and member of the pyOpenSci community. He's also a generally a great human being who is growing open science efforts such as 

Reviewer: ```suggestion
Ivan Ogasawara is a long time advisor, editor and member of the pyOpenSci community. He's also a generally a great human being who is growing open science efforts such as 
```

[Open Science Labs](https://opensciencelabs.org/); which is a gloabl community devoted to education efforts and tools that support open science.  

Ivan was quick to point out some basic metrics offered by GitHub which follow their 
[community standards online guidebook here](https://opensource.guide/).

<blockquote class="twitter-tweet"><p lang="en" dir="ltr">Maybe not totally related, but github has a section called community standards that mabe could be used as reference, for example: <a href="https://t.co/wmu1bDdcQR">https://t.co/wmu1bDdcQR</a></p>&mdash; XMN (@xmnlab) <a href="https://twitter.com/xmnlab/status/1578745372880429057?ref_src=twsrc%5Etfw">October 8, 2022</a></blockquote> <script async src="https://platform.twitter.com/widgets.js" charset="utf-8"></script> 

> Actually it's totally related, Ivan! Let's have a look look at the [pyOpenSci contributing-guide GitHub repository](https://www.github.com/pyopensci/contributing-guide/community) to see how we are 

Reviewer: This link should be peer-review-guide now right?
I couldn't find the page you meant to link to, though

Author: it's the right link if you just use peer-review instead of contributing. it is the community checks page in github that all repos have.

Author: are you not able to see it? i didn't think it needed permissions to view if it's an open repo

doing as an organization.

Note that we are missing some important components:

* A code of conduct
* A contributing file that helps people understand how to contribute 
* Issue template for people opening issues 
* Pull request templates to guide people through opening pull requests 
* Repository admins accepting content reporting 

{% include figure image_path="/images/python-software-health/healthy-software-repo-file-checks-github.png" alt="Image showing the community standards page in GitHub. You can see in the image we are missing several critical files including  a code of conduct file, a contributing file that helps people understand how to contribute to the guide and issue and pull request templates." caption="Here you can see the community page in GitHub for our contributing-guide repository. Note that we are missing several important items in the repo including a code of conduct file, a contributing file that helps people understand how to contribute to the guide and issue and pull request templates. HELP!" %}

> > Um.... we've got some real work to do, y'all on our guides and repos. We need to set a better example here don't we? We welcome help welcome if you are reading this and wanna contribute. Just sayin... 

Reviewer: Should we raise issues, "Add CoC", "Add contributing.md", "Add pull request templates"?  
I would just raise them but I'm not quite sure where you want them.  
Do we want a separate governance repo with the CoC and contributing that we can link to from every other repo?

Author: i'm not sure how this relates to this blog post david. it seems like a comment that belongs elsewhere. i'm just making the point that our repos don't have the files if you look at that link we are missing a bunch of green checks.

But these checks don't look at what's in that README, or how the issue templates
are designed to invite contributions that are useful to the maintainers (and that
guide new potential contributors). 

In short, GitHub checks are excellent but mostly exterior infrastructure focused.
They don't check content of those files and items. 

#### So what do content checks look like?

As Chris mentions below, things like having a clearly stated goal and intention,
likely articulated in the **README** file is a sign of a healthy project. This goal was
ideally developed prior to development beginning. Further, if well-written, 
it helps keep the scope of the project management. 

<blockquote class="twitter-tweet"><p lang="en" dir="ltr">To this point - i think that an example of a healthy project behavior is that it explicitly states its technical and organizational goals and intentions</p>&mdash; Chris Holdgraf (@choldgraf) <a href="https://twitter.com/choldgraf/status/1577971815305060354?ref_src=twsrc%5Etfw">October 6, 2022</a></blockquote> <script async src="https://platform.twitter.com/widgets.js" charset="utf-8"></script> 

> ### Test suits and Python versions

Reviewer: ```suggestion
### Test suites and Python versions
```
